### ISSUE TITLE
Weird behaviour with show of custom types

### ISSUE THREAD

**Issue by jakobnissen** [JuliaLang/julia#45776]
The docstring of `IOContext` says:
> :compact: Boolean specifying that values should be printed more compactly, e.g. that numbers should be printed with fewer digits. This is set when printing array elements.

That appears to be wrong:
```
julia> struct Foo end

julia> function Base.show(io::IO, ::MIME"text/plain", ::Foo)
           if get(io, :compact, false)
               print(io, "compact")
           else
               print(io, "verbose")
           end
       end;

julia> [Foo()]
1-element Vector{Foo}:
 verbose
```

And here is something even wilder: Calling `println` instead of `print` in the definition of `show` causes it to dispatch to a different method. I have zero idea what is going on here.
```
julia> struct Foo end

julia> Base.show(io::IO, ::MIME"text/plain", ::Foo) = print(io, "compact")

julia> [Foo()]
1-element Vector{Foo}:
 compact

julia> Base.show(io::IO, ::MIME"text/plain", ::Foo) = println(io, "compact") # nb: println instead of print

julia> [Foo()]
1-element Vector{Foo}:
 Foo()
```

Beside these two issues which I think are bugs, it's completely unclear to me how to define a custom verbose display for a new type, and not use this display when inside a container. So the documentation here could probably be improved

**Comment by JeffBezanson** [comment]
Yes we should change the documentation. We should say that it MIGHT be set when printing array elements, e.g. if the container type in question thinks it has little space per element. But that's entirely up to the container.

The `println` thing is happening because arrays attempt printing elements more verbosely, but if the output contains a line break it calls a different function instead since the line break would mess up the array display. I don't think you can reasonably expect to control how your type is printed inside a certain type of container, since again that is up to the container.
